src/pudl/helpers.py

```diff
@@ -1167,3 +1170,170 @@ def get_working_eia_dates():
+            pd.Timedelta(hours=1))
+    })
+
+    df = (
+        # merge in the hours for the calculation
+        df.merge(dates_to_hours, on=['report_date'])
+        # actually calculate capacity factor wooo!
+        .assign(
+            capacity_factor=lambda x: x.net_generation_mwh / \
+                (x.capacity_mw * x.hours)
+        )
+        # Replace unrealistic capacity factors with NaN
+        .pipe(
+            oob_to_nan,
+            ['capacity_factor'],
+            lb=min_cap_fact,
+            ub=max_cap_fact
+        )
+        .drop(['hours'], axis=1)
+    )
+    return df
+
+
+def weighted_average(df, data_col, weight_col, idx_cols):
+    """
+    Generate a weighted average.
+
+    Args:
+        df (pandas.DataFrame): A DataFrame containing, at minimum, the columns
+            specified in the other parameters data_col and weight_col.
+        data_col (string): column name of
+        weight_col (string): column name to weight on
+        idx_cols (list): A list of the columns to group by when calcuating
+            the weighted average value.
+
+    Returns:
+        pandas.DataFrame: a table with idx_cols and the weighted data_col.
+    """
+    df['_data_times_weight'] = df[data_col] * df[weight_col]
+    df['_weight_where_notnull'] = df[weight_col] * pd.notnull(df[data_col])
+    g = df.groupby(idx_cols, observed=True)
+    result = (
+        g['_data_times_weight'].sum(min_count=1)
+        / g['_weight_where_notnull'].sum(min_count=1)
+    )
+    del df['_data_times_weight'], df['_weight_where_notnull']
+    return result.to_frame(name=data_col).reset_index()
+
+
+def agg_cols(df_in, id_cols, sum_cols, wtavg_dict):
+    """
+    Aggregate dataframe by summing and using weighted averages.
+
+    Args:
+        df_in (pandas.DataFrame): input table to aggregate. Must have columns
+            in ``id_cols``, ``sum_cols`` and keys from ``wtavg_dict``.
+        id_cols (iterable): columns to group/aggregate based on. These columns
+            will be passed as an argument into grouby as ``by``.
+        sum_cols (iterable): columns to sum.
+        wtavg_dict (dictionary): dictionary of columns to average (keys) and
+            columns to weight by (values).
+
+    """
+    cols_to_grab = id_cols
+    cols_to_grab = list(
+        set([x for x in cols_to_grab if x in list(df_in.columns)]))
+    if set(cols_to_grab) != set(id_cols):
+        warn(f"um {[x for x in id_cols if x not in cols_to_grab]}")
+    logger.debug(f'grouping by {cols_to_grab}')
+
+    df_out = (
+        df_in.groupby(by=cols_to_grab, as_index=False, observed=True)
+        [sum_cols]
+        .sum(min_count=1)
+    )
+
+    for data_col, weight_col in wtavg_dict.items():
+        df_out = weighted_average(
+            df_in,
+            data_col=data_col,
+            weight_col=weight_col,
+            idx_cols=cols_to_grab
+        ).merge(df_out, how='outer', on=cols_to_grab)
```

Comment:
this is going to be TMI but here ya go.... so I tried three different types of methods to get the same effect here and using `loc` to assign the columns consistenty won the speed game. here is what I did:

```
def sum_and_weighted_average_df(
        df_in: pd.DataFrame,
        by: list,
        sum_cols: list,
        wtavg_dict: Dict[str, str],
        method: str
) -> pd.DataFrame:
    """
    Aggregate dataframe by summing and using weighted averages.

    Many times we want to aggreate a data table using the same groupby columns
    but with different aggregation methods. This function combines two of our
    most common aggregation methods (summing and applying a weighted average)
    into one function. Because pandas does not have a built-in weighted average
    method for groupby we use :func:``weighted_average``.

    Args:
        df_in (pandas.DataFrame): input table to aggregate. Must have columns
            in ``id_cols``, ``sum_cols`` and keys from ``wtavg_dict``.
        by (list): columns to group/aggregate based on. These columns
            will be passed as an argument into grouby as ``by``.
        sum_cols (list): columns to sum.
        wtavg_dict (dictionary): dictionary of columns to average (keys) and
            columns to weight by (values).

    Returns:
        table with join of columns from ``by``, ``sum_cols`` and keys of
        ``wtavg_dict``. Primary key of table will be ``by``.

    """
    logger.debug(f'grouping by {by}')
    as_index = False
    if method == 'loc_assign':
        as_index = True
    df_out = (
        df_in.groupby(by=by, as_index=as_index, observed=True)
        [sum_cols]
        .sum(min_count=1)
    )

    if method == 'concat':
        dfs = []
        for data_col, weight_col in wtavg_dict.items():
            dfs.append(weighted_average(
                df_in,
                data_col=data_col,
                weight_col=weight_col,
                idx_cols=by
            )
            )
        df_out = pd.merge(
            df_out,
            pd.concat(dfs, axis=1).reset_index(),
            how='outer',
            on=by
        )
    elif method == 'loc_assign':
        # df_out = df_out.set_index(by)
        for data_col, weight_col in wtavg_dict.items():
            df_out.loc[:, data_col] = weighted_average(
                df_in,
                data_col=data_col,
                weight_col=weight_col,
                idx_cols=by
            )[data_col]
        df_out = df_out.reset_index()
    else:
        for data_col, weight_col in wtavg_dict.items():
            df_out = pd.merge(
                df_out,
                weighted_average(
                    df_in,
                    data_col=data_col,
                    weight_col=weight_col,
                    idx_cols=by).reset_index(),
                how='outer',
                on=by
            )
    return df_out
```

and the notebook output consistently looked like this:
```
unit_part = PlantPart('plant_unit')

%%time
out1 = unit_part.ag_part_by_own_slice(gens_mega, method='concat')

begin aggregation for: plant_unit
CPU times: user 3.79 s, sys: 572 ms, total: 4.36 s
Wall time: 5.04 s

%%time
out2 = unit_part.ag_part_by_own_slice(gens_mega, method='merge')

begin aggregation for: plant_unit
CPU times: user 3.77 s, sys: 621 ms, total: 4.39 s
Wall time: 4.75 s

%%time
out3 = unit_part.ag_part_by_own_slice(gens_mega, method='loc_assign')

begin aggregation for: plant_unit
CPU times: user 3.59 s, sys: 579 ms, total: 4.17 s
Wall time: 4.51 s

```

(I also double checked with each output whether the dfs were compatible)

So imma gonna go with the loc assign method here, but lmk if you think otherwise.